File: algorithms/java/src/ValidParentheses.java
import java.util.Stack;

/**
 * @author jasper
 * @create 2021-10-12 13:34
 */
public class ValidParentheses {
    public boolean isValid(String s) {
        Stack<Character> brackets = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c){
                case '(':
                case '[':
                case '{':
                    brackets.push(c);
                    break;
                case ')':
                    if (!brackets.empty() && brackets.peek() == '('){
                        brackets.pop();
                    }else {
                        return false;
                    }
                    break;
                case ']':
                    if (!brackets.empty() && brackets.peek() == '['){
                        brackets.pop();
                    }else {
                        return false;
                    }
                    break;
                case '}':
                    if (!brackets.empty() && brackets.peek() == '{'){
                        brackets.pop();
                    }else {
                        return false;
                    }
                    break;
            }
        }
        return brackets.empty();
    }
}

File: algorithms/python/ValidParentheses.py
# @author jasper
# @create 2021-10-12 13:34


class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for t in s:
            if t == ')':
                try:
                    current = stack.pop()
                    if current != '(':
                        return False
                except:
                    return False
            elif t == '}':
                try:
                    current = stack.pop()
                    if current != '{':
                        return False
                except:
                    return False
            elif t == ']':
                try:
                    current = stack.pop()
                    if current != '[':
                        return False
                except:
                    return False
            else:
                stack.append(t)
        if len(stack) == 0:
            return True
        else:
            return False
